wordpress-mobile/WordPress-Android
# [Save Post Task Force] Prevent auto-saves after a user-initiated save and remove `SaveOnEditorExit` exception logging

Internal Reference: pb3aDo-18j-p2

## Description

This PR adds a flag to prevent triggering “**auto-save**” events after a “**user-initiated**” save event.

This is because the “**user-initiated**”/ “**save on exit**” event captures the latest state of the UI (the source of truth) and any further updates received from `Gutenberg` after this point (and after [the title and content serialization fix](https://github.com/WordPress/gutenberg/pull/36072)) will represent the same state.

We can now remove the [SaveOnExitException](https://github.com/wordpress-mobile/WordPress-Android/blob/71240284d0f0d3aeb16a79416f45a394f7585e2b/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L1763-L1775) logging in the `DEBUG` build because:

1. The ”**auto-save**”  events that get the content from the UI/ Gutenberg editor (see T1-T4) are queued as they're [separated by delays](https://github.com/wordpress-mobile/WordPress-Android/blob/71240284d0f0d3aeb16a79416f45a394f7585e2b/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/StorePostViewModel.kt#L100); a “**user-initiated**” save (S) event may complete before the queue is empty. All pending “**auto-saves**” (see T3-T4) after  “**user-initiated**” save are prevented from starting by adding the flag above. This makes the [AutosavePending](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/ui/posts/SaveOnExitException.kt#L13-L17) condition redundant as we’re preventing any further “auto-saves”.

2. The “user-initiated” save (S) event can update the `PostModel` if a queued “auto-save” (T2) for a previous UI state completes before it. In this case, [NoAutosavePending](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/ui/posts/SaveOnExitException.kt#L23-L26) becomes redundant as the “user-initiated” save (S) had changes to be saved after “auto-save” (T2) while there was no auto-save pending due to 1 above.

   <img width="430" alt="thread-state" src="https://user-images.githubusercontent.com/1405144/139619067-f9c69e08-d1d6-4ba0-86c4-c405591f0993.png">

~~Note: This PR is still in draft as it needs to target https://github.com/wordpress-mobile/WordPress-Android/pull/15508 which is currently failing + I had local `gutenberg-mobile` issues so could not test the final changes~~.  

## Testing Instructions

### Prerequisites

1. Create and publish a long post (copy content from this [txt file](https://cloudup.com/iX9yC9byRyN)) in one of your test sites using Calypso.

2. Check out this branch.

3. Apply [this patch](https://cloudup.com/iA4hNRd2XBe) to add logs.

5. Install and run the build.

### To Test

**Case 1: Published Post + **Save on Exit** using **Update** button** 

1. Open the long post created above in the Gutenberg editor.
2. Change the title of the post by typing a few letters (make sure you remember the letters in the new title)
3. Quickly click the **Update** button on the editor.
4. Wait for the editor to exit and the post to be updated.
5. Notice in the logs that:
	a. [AutosavePending](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/ui/posts/SaveOnExitException.kt#L13-L17) exception (if gets logged) is a case of a false positive as we're preventing "auto-saves" after "user-initiated" save.
	b. [NoAutosavePending](https://github.com/wordpress-mobile/WordPress-Android/blob/70ce7de8817020d29768074bb408688436320220/WordPress/src/main/java/org/wordpress/android/ui/posts/SaveOnExitException.kt#L23-L26) exception (if gets logged) is a case of a false positive as the  “user-initiated” save actually had changes to be saved after a previous “auto-save” state. (See [previously ran test results](https://cloudup.com/iNf7Khi8bJm) with matched hash-code logs.)

Repeat same tests for the below cases: 

Case 2:  Published Post + **Save on Exit** using **Back** button 
Case 3:  Draft post + **Save on Exit** using **Publish** button 
Case 4:  Draft post + **Save on Exit** using **Back** button 

## Review/ Merge Instructions

Only 1 reviewer is sufficient.

1. ~~Make sure https://github.com/wordpress-mobile/WordPress-Android/pull/15508 is merged to develop.~~
2. Review this PR.
3. Merge  [issue/15509-remove-save-on-exit-exception](https://github.com/wordpress-mobile/WordPress-Android/tree/issue/15509-remove-save-on-exit-exception) and review changes.
4. Remove "Not Ready for Merge" label.
5. Merge the PR.

## Regression Notes
1. Potential unintended areas of impact
All flows which save a post on Editor exit.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested different cases manually (see above). 

3. What automated tests I added (or what prevented me from doing so)
None

PR submission checklist:
